src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/Marshal.Unix.cs

```diff
@@ -154,6 +154,41 @@ internal static unsafe IntPtr AllocBSTR(int length)
             return s;
         }
 
+        internal static unsafe IntPtr SysAllocStringByteLen(byte[]? str, uint length)
+        {
+            // SysAllocString on Windows aligns the memory block size up
+            const nuint WIN32_ALLOC_ALIGN = 15;
+
+            ulong cbNative = (ulong)(uint)length + (uint)sizeof(IntPtr) + (uint)sizeof(char) + WIN32_ALLOC_ALIGN;
+            if (cbNative > uint.MaxValue)
+            {
+                throw new OutOfMemoryException();
+            }
+
+            IntPtr p = Interop.Sys.MemAlloc((nuint)cbNative & ~WIN32_ALLOC_ALIGN);
+            if (p == IntPtr.Zero)
+            {
+                throw new OutOfMemoryException();
+            }
+
+            IntPtr s = p + sizeof(IntPtr);
+            *(((uint*)s) - 1) = (uint)length;
+
+            if (str is not null)
+            {
+                fixed (byte* firstChar = str)
+                {
+                    Buffer.Memmove((byte*)s, firstChar, length);
+                }
+            }
+
+            // NULL-terminate with both a narrow and wide zero.
+            *(byte*)((byte*)s + length) = (byte)'\0';
+            *(short*)((byte*)s + ((length + 1) & ~1)) = 0;
```

Comment:
Can this lead to a buffer overrun?

Reply:
Unlikely as this is now matching the native implementation. I have verified by printing bytes before and after making sure there was no diff. Also, the CI is green.